### PR TITLE
scanner: Allow full throttle if there is no parallel disk ops

### PR DESCRIPTION
**Pull request by vadmeste**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
When there is no ongoing xl storage operations, it is okay to allow the scanner to go
on full speed scan.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/xl-storage-disk-id-check.go [hunk 1]
```diff
@@ -665,10 +666,21 @@ func scannerTrace(s scannerMetric, startTime time.Time, duration time.Duration,
 }
 
 // Update storage metrics
-func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...string) func(err *error) {
+func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, ctx context.Context, paths ...string) func(err *error) {
 	startTime := time.Now()
 	trace := globalTrace.NumSubscribers(madmin.TraceStorage) > 0
+	incS3Count := false
+
+	if tc, ok := ctx.Value(mcontext.ContextTraceKey).(*mcontext.TraceCtxt); ok && strings.HasPrefix(tc.FuncName, "s3.") {
```
Comment (vadmeste) [inline cmd/xl-storage-disk-id-check.go]: >  strings.HasPrefix(tc.FuncName, "s3.")

This doesn't look robust enough for now; maybe we need to set a new context value that can be spread between nodes using another header

File: cmd/xl-storage-disk-id-check.go [hunk 2]
```diff
@@ -665,10 +666,21 @@ func scannerTrace(s scannerMetric, startTime time.Time, duration time.Duration,
 }
 
 // Update storage metrics
-func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, paths ...string) func(err *error) {
+func (p *xlStorageDiskIDCheck) updateStorageMetrics(s storageMetric, ctx context.Context, paths ...string) func(err *error) {
```
Comment (harshavardhana) [inline cmd/xl-storage-disk-id-check.go]: context must be first argument @vadmeste

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by klauspost** [comment]
The `diskHealthTracker`  in `xlStorageDiskIDCheck` already keeps track of running operations - so when `cap(diskHealthTracker.tokens) == ` len(diskHealthTracker.tokens)` there are no ongoing disk activity.  The scanner is already excluded from this.

QUESTION is if we really want to do that by default?

File: cmd/xl-storage-disk-id-check.go (hunk 1 above)
Comment (klauspost) [reply cmd/xl-storage-disk-id-check.go]: Not needed. The health tracker already tracks ongoing operations and there is no magic there.

**Comment by vadmeste** [comment]
> QUESTION is if we really want to do that by default?

The main goal is to make scanning goes in full throttle when we know that there is no ongoing S3 operations; Since one S3 operation involves multiple disks in different nodes, I thought the easiest way is to keep track of the number of concurrent S3 operations per disk using context.Context concept; (this seems the cheapest option in my opinion)

So, if an S3 request lands one node; we add an attribute to context.Context which will be passed down to xl-storage, that way we can count the number of concurrent s3 ops per disk. To make sure that remote disks get the same attribute, we can pass that in internode calls headers; 

Now each disk that is currently scanning knows how many concurrent requests that is hitting and dynamically decides if it should go in full throtlle or uses the usual sleeping mechanism

**Comment by klauspost** [comment]
> The main goal is to make scanning goes in full throttle when we know that there is no ongoing S3 operations

Yes. But I am not too sure everybody will appreciate the extra wear on the hardware.

> (monitor) S3 operations

It is a hacky, fragile solution. When we already monitor running disk operations we should hook into that instead.

You can add a parameter in `(p *xlStorageDiskIDCheck) NSScanner(..)` that passes down a function on whether wait should be skipped.

**Comment by vadmeste** [comment]
> Yes. But I am not too sure everybody will appreciate the extra wear on the hardware.

Yes, maybe let's discuss this first; what we can possibly do is to avoid running the scanner when we know there is no write s3 operations since the last scanner run; I actually started to believe we need to merge the other scanner PR first, https://github.com/minio/minio/pull/18107 ; I have few left things to check before making it ready for review

File: cmd/data-scanner.go [hunk 3]
```diff
@@ -246,6 +246,8 @@ type folderScanner struct {
 	healObjectSelect      uint32 // Do a heal check on an object once every n cycles. Must divide into healFolderInclude
 	scanMode              madmin.HealScanMode
 
+	fullThrottle func() bool
```
Comment (harshavardhana) [inline cmd/data-scanner.go]:
```suggestion
	weSleep func() bool
```

is better @vadmeste which returns `true` when `cap(tokens) != len(tokens)`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/data-scanner.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/data-scanner.go]: PTAL @vadmeste

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
I have rebased and pushed the changes @vadmeste

**Review by shtripat — APPROVED** [review]
lgtm

**Comment by harshavardhana** [comment]
This PR is merged, @vadmeste will send another PR to allow this to be configurable behavior via `scanner speed=default delay=always`

**Pull request merged**